# Pulling Stock Data to Analyze Companies with yfinance Library
### Using yfinance and pandas to pull stock data and compare across multiple tickers

+ First step: do pip install of pandas and yfinance libraries, if you haven't installed them before.  
+  In Jupyter, these are already available and installed, so I've commented them out, just in case you need them for your system.    
+  This first code block defines the stock to analyze and creates a stock object.

In [49]:
# !pip install pandas
# !pip install yfinance

# Import necessary libraries
import pandas as pd
import yfinance as yf

# Define the stock symbol to research
symbol = "PFE"
stock = yf.Ticker(symbol)

print("stock is: " + str(type(stock)))

stock is: <class 'yfinance.ticker.Ticker'>


### Let's take a look at the huge variety of free data yfinance lets us pull  
+ As you can see, it appears a bit overwhelming.
+ Don't worry too much. You don't have to use it all. Just skim and see what interests you.

In [5]:
# Display all data for the stock object, which is in dictionary form
stock.info

{'address1': '66 Hudson Boulevard East',
 'city': 'New York',
 'state': 'NY',
 'zip': '10001-2192',
 'country': 'United States',
 'phone': '212 733 2323',
 'website': 'https://www.pfizer.com',
 'industry': 'Drug Manufacturers - General',
 'industryKey': 'drug-manufacturers-general',
 'industryDisp': 'Drug Manufacturers - General',
 'sector': 'Healthcare',
 'sectorKey': 'healthcare',
 'sectorDisp': 'Healthcare',
 'longBusinessSummary': 'Pfizer Inc. discovers, develops, manufactures, markets, distributes, and sells biopharmaceutical products in the United States and internationally. The company offers medicines and vaccines in various therapeutic areas, including cardiovascular and migraine under the Eliquis, Nurtec ODT/Vydura, Zavzpret, and the Premarin family brands; infectious diseases with unmet medical needs under the Prevnar family, Abrysvo, Nimenrix, FSME/IMMUN-TicoVac, and Trumenba brands; and COVID-19 prevention and treatment, and potential future mRNA and antiviral products und

### Ok, let's cherry-pick a few key pieces of data to display from the dictionary, to get used to working with it

In [68]:
# Use the get method to store and display a few key ratios and data points from the dictionary
symbol = "PFE"
market_price = stock.info.get('currentPrice')
market_cap = stock.info.get('marketCap')
trailing_pe = stock.info.get('trailingPE')
forward_pe = stock.info.get('forwardPE')
beta = stock.info.get('beta')
div_yield = stock.info.get('trailingAnnualDividendYield')
div_po=stock.info.get('payoutRatio')
de_ratio = stock.info.get('debtToEquity')
roa = stock.info.get('returnOnAssets')
roe = stock.info.get('returnOnEquity')
fcf = stock.info.get('freeCashflow')
cr = stock.info.get('currentRatio')
qr = stock.info.get('quickRatio')

# Notice the use of f-strings to format the output, since the data is type float
print("Symbol: " + symbol)
print(f"Market Price: ${market_price:,.2f}") 
print(f"Market Cap: ${market_cap/1000000000:,.3f} B") 
print(f"Trailing PE: {trailing_pe:,.2f}")
print(f"Forward PE: {forward_pe:,.2f}")
print(f"Beta: {beta:,.2f}")
print(f"Div. Yield: {div_yield * 100:,.2f}%")
print(f"Payout Ratio: {div_po * 100:,.2f}%")
print(f"D/E Ratio: {de_ratio:,.1f}%")
print(f"ROA: {roa * 100:,.2f}%")
print(f"ROE: {roe * 100:,.2f}%")
print(f"FCF: {fcf/1000000000:,.3f} B")
print(f"Current Ratio: {cr:,.2f}")
print(f"Quick Ratio: {qr:,.2f}")

Symbol: PFE
Market Price: $23.98
Market Cap: $136.396 B
Trailing PE: 12.69
Forward PE: 8.19
Beta: 0.46
Div. Yield: 7.09%
Payout Ratio: 89.95%
D/E Ratio: 69.7%
ROA: 5.25%
ROE: 12.17%
FCF: 16.018 B
Current Ratio: 1.16
Quick Ratio: 0.78


### As you can see, it's quick and easy to get lots of useful data to compare across companies, all in one place
+ For example, to analyze this we would have had to scroll three or four different financial statements
+ (e.g. Statement of Cashflows for FCF, Balance Sheet for QR and CR) 

### Additionally we can already make some basic inferences and analyses about the company
+ For example, the liquidity ratios of 1.16 and 0.78 are a bit low (compared to standard CR of 1.5 to 2.0 and QR above 1.0)
+ This could indicate a liquidity crunch if cash flows dry up.
+ However, we also see the most recent free cash flow was significant (over 16 billion) against a market cap of 136 billion
+ This is a price to free cash flow of 8.5, which is very solid. So while liquidity is down, there is plenty of cash to pay bills, dividends, do capex
+ Beta of 0.46 suggests less volatile returns than the market portfolio
+ Dividend yield is pretty high (above 7%) yet payout ratio is nearly 90%. Sustainability of that dividend may be questionable.
+ Financial leverage is also significant, with D/E ratio at 70%
+ ROA is pretty low, suggesting efficiency in generating profits from assets may not be up to par. We should compare  similar companies to be sure.
+ Investors are pricing the stock as such:
+ Very low trailing PE and forward PE in the 8 to 13 range implies investors expect very limited growth prospects in the future.

#### Note: these data points and analyses will undoubtedly be different by the time you see this! All these KPIs change frequently

### As you can see, this is just a small amount of data gleaned from ONE stock.
+ We were able to do quite a bit of useful analysis in a very short time
+ Play around with the above code. Change the symbol out for your favorite ticker.
+ With no GUI and not even a dataframe yet, we can ALREADY do quite a bit of useful analysis.

### Next up: let's make a dataframe to compare these key data points across multiple companies in the industry